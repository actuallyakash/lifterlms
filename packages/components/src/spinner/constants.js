/**
 * Spinner wrapper element class name.
 *
 * @type {string}
 */
export const WRAPPER_CLASSNAME = 'llms-spinning';

/**
 * Spinner element class name.
 *
 * @type {string}
 */
export const CLASSNAME = 'llms-spinner';

/**
 * Small-sized spinner identifier / class name.
 *
 * @type {string}
 */
export const SIZE_SMALL = 'small';

/**
 * Default-sized spinner identifier / class name.
 *
 * @type {string}
 */
export const SIZE_DEFAULT = 'default';
