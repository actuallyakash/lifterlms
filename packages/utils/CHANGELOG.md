@lifterlms/utils CHANGELOG
==========================

v1.0.0 - 2022-03-08
-------------------

+ Initial public release.
