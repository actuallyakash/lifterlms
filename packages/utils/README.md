# LifterLMS JS Utilities

Javascript utility library for use in LifterLMS and LifterLMS add-ons.

## Usage in a LifterLMS add-on

This package is included in the LifterLMS core plugin as a module and registered using the WordPress scripts dependency API (see [wp_register_script()][https://developer.wordpress.org/reference/functions/wp_register_script/]).

To use components you can add `llms-utils` as a dependency to your script and access the module via `window.llms.components`, for example:

```js
const { getAdminUrl } = window.llms.utils;
```

## Installation

Install the module as a dependency in your project:

```bash
npm i -S @lifterlms/utils`
```

## LifterLMS Versions

The following table records the module version and which LifterLMS versions it has been included in.

| Module Version | LifterLMS Version |
| -------------- | ----------------- |
| 1.0.0          | 6.0.0             |


## Changelog

[View the Changelog](./CHANGELOG.md)

## API Docs

<!-- START TOKEN(Autogenerated API docs) -->

### getAdminUrl

Retrieves the WordPress admin URL.

This function relies on the presence of localized data from the LifterLMS plugin which is only
present on the WordPress admin panel. If used out of context a default partial path url, `/wp-admin`
will be returned.

_Returns_

-   `string`: The WP Admin URL.

### trailingSlashIt

Adds a trailing forward slash to a given string.

_Parameters_

-   _str_ `string`: A string with or without a trailing forward slash.

_Returns_

-   `string`: The original string with a trailing forward slash added.

### untrailingSlashIt

Remove trailing forward slash from a given string.

_Parameters_

-   _str_ `string`: A string with or without a trailing forward slash.

_Returns_

-   `string`: The original string with the trailing forward slash removed.


<!-- END TOKEN(Autogenerated API docs) -->
