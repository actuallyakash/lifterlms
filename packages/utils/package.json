{
  "name": "@lifterlms/utils",
  "version": "1.0.1-alpha.1",
  "description": "JS Utility library for LifterLMS.",
  "author": "Team LifterLMS <dev@lifterlms.com>",
  "license": "GPL-3.0-or-later",
  "homepage": "https://github.com/gocodebox/lifterlms/tree/trunk/packages/utils",
  "keywords": [
    "lifterlms",
    "wordpress",
    "utils",
    "ui"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/gocodebox/lifterlms.git",
    "directory": "packages/utils"
  },
  "bugs": {
    "url": "https://github.com/gocodebox/lifterlms/labels/package%3A%utils"
  },
  "main": "src/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "docgen": "docgen src/index.js --output README.md --to-token",
    "dev": "./../dev/src/index.js",
    "lint:js": "wp-scripts lint-js ./ --config ../../.eslintrc.js",
    "test": "wp-scripts test-unit-js ./ --config ../scripts/config/jest-unit.config.js --verbose"
  }
}
