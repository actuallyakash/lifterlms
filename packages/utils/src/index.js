export * from './formatting';
export * from './url';
