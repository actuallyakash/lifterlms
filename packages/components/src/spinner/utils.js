// WP deps.
import { __ } from '@wordpress/i18n';

// Internal deps.
import { WRAPPER_CLASSNAME, CLASSNAME, SIZE_DEFAULT } from './constants';
import { STYLES } from './styles';

/**
 * Creates a spinner element inside the specified wrapper Element.
 *
 * @since 1.1.0
 *
 * @param {Element} wrapper DOM node to append the created spinner element to.
 * @param {string}  size    Spinner element size.
 * @return {Element} Returns the created spinner node.
 */
export function create( wrapper, size = SIZE_DEFAULT ) {
	const spinner = document.createElement( 'div' ),
		loadingMsg = __( 'Loadingâ€¦', 'lifterlms' );

	spinner.innerHTML = `<i class="${ CLASSNAME } ${ size }" role="alert" aria-live="assertive"><span class="screen-reader-text">${ loadingMsg }</span></i>`;
	spinner.classList.add( WRAPPER_CLASSNAME );

	wrapper.appendChild( spinner );

	return spinner;
}

/**
 * Normalizes accepted selector inputs and returns a `NodeList`.
 *
 * When jQuery selection is detected, adds a console deprecation warning as well.
 *
 * @since 1.1.0
 *
 * @param {NodeList|Element|string|jQuery} selector The input selector.
 * @return {Element[]} An array of `Element` objects derived from the selector input.
 */
export function ensureElementList( selector ) {
	selector =
		typeof selector === 'string'
			? document.querySelectorAll( selector )
			: selector;

	// Already a NodeList.
	if ( selector instanceof NodeList ) {
		return Array.from( selector );
	}

	const list = [];
	if ( selector instanceof Element ) {
		list.push( selector );
	} else if ( typeof jQuery !== 'undefined' && selector instanceof jQuery ) {
		selector.toArray().forEach( ( el ) => list.push( el ) );
	}

	return list;
}

/**
 * Locates an existing spinner element which is a direct child of the specified wrapper element.
 *
 * @since 1.1.0
 *
 * @param {Element} wrapper Node element for the wrapper.
 * @return {null|undefined|Element} Returns `null` if no spinners exist within the wrapper, undefined if no spinners are
 *                                  direct descendants of the wrapper, otherwise returns the spinner element.
 */
export function find( wrapper ) {
	const spinners = wrapper.querySelectorAll( `.${ WRAPPER_CLASSNAME }` );
	if ( ! spinners.length ) {
		return null;
	}

	return Array.from( spinners ).find( ( el ) => wrapper === el.parentNode );
}

/**
 * Loads CSS styles and appends them to the document's <head>.
 *
 * Attaching CSS directly to the `get()` method means that we don't have to worry about loading CSS files (or relying on CSS) included
 * by the LifterLMS core plugin in order to use this styled component.
 *
 * @since 1.1.0
 *
 * @return {void}
 */
export function loadStyles() {
	const STYLE_ID = 'llms-spinner-styles';

	if ( ! document.getElementById( STYLE_ID ) ) {
		const style = document.createElement( 'style' );
		style.textContent = STYLES.replace( /\n/g, '' )
			.replace( /\t/g, ' ' )
			.replace( /\s\s+/g, ' ' );
		style.id = STYLE_ID;
		document.head.appendChild( style );
	}
}
