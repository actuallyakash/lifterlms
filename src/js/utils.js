import * as utils from '../../packages/utils/src';

/**
 * Expose @lifterlms/utils via the global `window.llms` object.
 *
 * @since 6.0.0
 */
window.llms = window.llms || {};
window.llms.utils = utils;
