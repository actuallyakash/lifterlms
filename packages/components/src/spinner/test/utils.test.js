import { fakeQuery } from './__MOCKS__/jquery';

import { create, ensureElementList, find, loadStyles } from '../utils';
import { SIZE_SMALL, SIZE_DEFAULT, WRAPPER_CLASSNAME } from '../constants';

global.jQuery = fakeQuery;

describe( 'Spinner: utils', () => {
	describe( 'create()', () => {
		const tests = [
			{
				name: 'Default size (not specified)',
				size: undefined,
			},
			{
				name: 'Default size (passed)',
				size: SIZE_DEFAULT,
			},
			{
				name: 'Small size',
				size: SIZE_SMALL,
			},
		];

		test.each( tests )( '$name', ( { size } ) => {
			document.body.innerHTML = '<div id="wrapper"></div>';
			const el = create( document.querySelector( '#wrapper' ), size );
			expect( el ).toBeInstanceOf( Element );
			expect( el.innerHTML ).toMatchSnapshot();
		} );
	} );

	describe( 'ensureElementList()', () => {
		const tests = [
			{
				name: 'NodeList',
				getInput: () => document.querySelectorAll( 'div' ),
			},
			{
				name: 'Element',
				getInput: () => document.querySelector( '.abc' ),
			},
			{
				name: 'String (single)',
				getInput: () => '.def',
			},
			{
				name: 'String (multiple)',
				getInput: () => 'div',
			},
			{
				name: 'jQuery',
				getInput: () => global.jQuery( 'div' ),
			},
		];

		test.each( tests )( '$name input', ( { getInput } ) => {
			document.body.innerHTML = `
				<div class="abc"></div>
				<div class="def"></div>
			`;

			const input = getInput(),
				list = ensureElementList( input );

			expect( list ).toMatchSnapshot();
			list.forEach( ( el ) => {
				expect( el ).toBeInstanceOf( Element );
			} );

			// Make sure Elements aren't removed from the DOM.
			expect( document.body.innerHTML ).toMatchSnapshot();
		} );
	} );

	describe( 'find()', () => {
		test( 'No spinners found in wrapper', () => {
			document.body.innerHTML = `
				<div id="wrap">
					<div class="abc"></div>
				</div>
			`;

			expect( find( document.getElementById( 'wrap' ) ) ).toBeNull();
		} );

		test( 'No spinners that are direct children of the wrapper', () => {
			document.body.innerHTML = `
				<div id="wrap">
					<div class="abc"><div class="${ WRAPPER_CLASSNAME }"></div></div>
				</div>
			`;

			expect( find( document.getElementById( 'wrap' ) ) ).toBeUndefined();
		} );

		test( 'Spinner found', () => {
			document.body.innerHTML = `
				<div id="wrap">
					<div class="${ WRAPPER_CLASSNAME }" id="shouldbefound"></div>
				</div>
			`;

			const spinner = find( document.getElementById( 'wrap' ) );
			expect( spinner ).toBeInstanceOf( Element );
			expect( spinner.id ).toBe( 'shouldbefound' );
		} );
	} );

	test( 'loadStyles()', () => {
		document.head.innerHTML = '';

		// Load them.
		loadStyles();
		const headHTML = document.head.innerHTML;
		expect( headHTML ).toMatchSnapshot();

		// Doesn't load them again.
		loadStyles();
		expect( document.head.innerHTML ).toBe( headHTML );
	} );
} );
