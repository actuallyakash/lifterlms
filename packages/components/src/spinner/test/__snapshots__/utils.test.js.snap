// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner: utils create() Default size (not specified) 1`] = `"<i class=\\"llms-spinner default\\" role=\\"alert\\" aria-live=\\"assertive\\"><span class=\\"screen-reader-text\\">Loading…</span></i>"`;

exports[`Spinner: utils create() Default size (passed) 1`] = `"<i class=\\"llms-spinner default\\" role=\\"alert\\" aria-live=\\"assertive\\"><span class=\\"screen-reader-text\\">Loading…</span></i>"`;

exports[`Spinner: utils create() Small size 1`] = `"<i class=\\"llms-spinner small\\" role=\\"alert\\" aria-live=\\"assertive\\"><span class=\\"screen-reader-text\\">Loading…</span></i>"`;

exports[`Spinner: utils ensureElementList() Element input 1`] = `
Array [
  <div
    class="abc"
  />,
]
`;

exports[`Spinner: utils ensureElementList() Element input 2`] = `
"
				<div class=\\"abc\\"></div>
				<div class=\\"def\\"></div>
			"
`;

exports[`Spinner: utils ensureElementList() NodeList input 1`] = `
Array [
  <div
    class="abc"
  />,
  <div
    class="def"
  />,
]
`;

exports[`Spinner: utils ensureElementList() NodeList input 2`] = `
"
				<div class=\\"abc\\"></div>
				<div class=\\"def\\"></div>
			"
`;

exports[`Spinner: utils ensureElementList() String (multiple) input 1`] = `
Array [
  <div
    class="abc"
  />,
  <div
    class="def"
  />,
]
`;

exports[`Spinner: utils ensureElementList() String (multiple) input 2`] = `
"
				<div class=\\"abc\\"></div>
				<div class=\\"def\\"></div>
			"
`;

exports[`Spinner: utils ensureElementList() String (single) input 1`] = `
Array [
  <div
    class="def"
  />,
]
`;

exports[`Spinner: utils ensureElementList() String (single) input 2`] = `
"
				<div class=\\"abc\\"></div>
				<div class=\\"def\\"></div>
			"
`;

exports[`Spinner: utils ensureElementList() jQuery input 1`] = `
Array [
  <div
    class="abc"
  />,
  <div
    class="def"
  />,
]
`;

exports[`Spinner: utils ensureElementList() jQuery input 2`] = `
"
				<div class=\\"abc\\"></div>
				<div class=\\"def\\"></div>
			"
`;

exports[`Spinner: utils loadStyles() 1`] = `"<style id=\\"llms-spinner-styles\\"> .llms-spinning { background: rgba( 250, 250, 250, 0.7 ); bottom: 0; display: none; left: 0; position: absolute; right: 0; top: 0; z-index: 2; } .llms-spinner { animation: llms-spinning 1.5s linear infinite; box-sizing: border-box; border: 4px solid #313131; border-radius: 50%; height: 40px; left: 50%; margin-left: -20px; margin-top: -20px; position: absolute; top: 50%; width: 40px; } .llms-spinner.small { border-width: 2px; height: 20px; margin-left: -10px; margin-top: -10px; width: 20px; } @keyframes llms-spinning { 0% { transform: rotate( 0deg ) } 50% { border-radius: 5%; } 100% { transform: rotate( 220deg) } }</style>"`;
